### PR TITLE
Series implementation of auxiliary gamma function

### LINKED ISSUE #116: Performance bottleneck identified in lower incomplete gamma function
@AlexanderMath identified that much of the time in evaluating `eri_basis_sparse` is spent evaluating the lower incomplete gamma function here:

https://github.com/graphcore-research/pyscf-ipu/blob/023e11487762731117d73d14bc98b2f74226f800/pyscf_ipu/experimental/special.py#L95

### PR DESCRIPTION
**Pull request by hatemhelal**
This PR adds another implementation `gammanu` to make use the series expansion of the lower incomplete gamma function.  This is derived in the notebook included in this PR. 

Closes #116 which motivated this journey into obscure numerical methods.

Using the command `pytest --durations=8` will report the 8 slowest testpoints:

### CPU tests with `gammanu = gammanu_gamma`
```bash
================================================== slowest 8 durations ==================================================
39.20s call     test/test_experimental.py::test_water_eri[True]
34.44s call     test/test_experimental.py::test_water_eri[False]
7.30s call     test/test_experimental.py::test_eri
5.68s call     test/test_experimental.py::test_eri_basis
4.15s call     test/test_experimental.py::test_gto[6-31+g]
3.30s call     test/test_experimental.py::test_gto[sto-3g]
3.04s call     test/test_experimental.py::test_nuclear
2.29s call     test/test_experimental.py::test_water_nuclear
=========================== 22 passed, 4 skipped, 1 xfailed, 5 warnings in 112.22s (0:01:52) ============================
```

### CPU tests with `gammanu = gammanu_series`
```bash
================================================== slowest 8 durations ==================================================
10.05s call     test/test_experimental.py::test_eri
5.93s call     test/test_experimental.py::test_water_eri[True]
5.23s call     test/test_experimental.py::test_nuclear
4.84s call     test/test_experimental.py::test_eri_basis
4.17s call     test/test_experimental.py::test_gto[6-31+g]
3.47s call     test/test_experimental.py::test_gto[sto-3g]
2.41s call     test/test_experimental.py::test_water_nuclear
2.05s call     test/test_experimental.py::test_kinetic
================================= 22 passed, 4 skipped, 1 xfailed, 5 warnings in 50.32s =================================
```

### CODE REVIEW

File: notebooks/gammanu.ipynb [hunk 1]
```diff
@@ -0,0 +1,1105 @@
+{
```
Comment (awf) [inline notebooks/gammanu.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    gammanu = simplify(F_nu.doit())
</pre><p>Maybe call this <code>F_nu_expanded</code>? The name <code>gammanu</code> sounds like another of the gamma functions   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/graphcore-research/pyscf-ipu/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (awf) [inline notebooks/gammanu.ipynb:1]: <p>"Some implementations will fail at t=0, so let's re-evaluate the integral for that special case:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/graphcore-research/pyscf-ipu/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (awf) [inline notebooks/gammanu.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    F_nu.subs(t, 0).doit()
<code>F_nu_expanded_t0 =F_nu.subs(t, 0).doit()</code>
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/graphcore-research/pyscf-ipu/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (awf) [inline notebooks/gammanu.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    help(lambdify((nu, t), gammanu, modules="scipy"))
</pre><p>Can we make it name it <code>F_nu_lambdifygenerated</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/graphcore-research/pyscf-ipu/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (awf) [inline notebooks/gammanu.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #15.    def gamma_nu(nu, t, epsilon=1e-10):
</pre><p>Again maybe rename F_nu_lambdifygenerated</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/graphcore-research/pyscf-ipu/pull/119/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (hatemhelal) [reply notebooks/gammanu.ipynb:1]: <p>not that I can easily see in the lambdify doc.</p>

File: pyscf_ipu/experimental/special.py [hunk 2]
```diff
@@ -77,23 +77,20 @@ def binom_lookup(x: IntN, y: IntN, nmax: int = LMAX) -> IntN:
 binom = binom_lookup
 
 
-def gammanu(nu: IntN, t: FloatN, epsilon: float = 1e-10) -> FloatN:
+def gammanu(nu: IntN, t: FloatN, num_terms: int = 128) -> FloatN:
```
Comment (awf) [inline pyscf_ipu/experimental/special.py:80]: Maybe leave both in, called e.g. `gammanu_series` and `gammnu_gamma`, and switch between them as you do with binom.  Then the PR can check timings with a clean switch within the PR?

File: test/test_experimental.py [hunk 3]
```diff
@@ -171,20 +157,8 @@ def test_eri():
...
-    for ijkl, expect in zip(indices, expected):
-        actual = eri_orbitals([basis.orbitals[aoid] for aoid in ijkl])
-        assert_allclose(actual, expect, atol=1e-4)
-
-
-def test_eri_basis():
-    # H2 molecule in sto-3g: See equation 3.235 of Szabo and Ostlund
-    h2 = molecule("h2")
-    basis = basisset(h2, "sto-3g")
-
```
Comment (awf) [inline test/test_experimental.py:186]: These deletions seem a bit orthogonal, and will contribute to test time?

Comment (hatemhelal) [reply test/test_experimental.py:186]: Good point, I think I'd like to reorganise the tests to break apart the growing file `test_experimental.py`

I started doing this on another branch but I think it would be cleaner to do this kind of reorg separately from other source level changes.  I can revert these changes and open a new issue to organise the tests (including removing this duplicated test)

Does that seem reasonable?

**Comment by hatemhelal** [comment]
I didn't spot this before but it looks like from the test timing report copied in the PR summary that the nuclear attraction integrals get slower with the switch to the series implementation of `gammanu`

**Comment by awf** [comment]
> I didn't spot this before but it looks like from the test timing report copied in the PR summary that the nuclear attraction integrals get slower with the switch to the series implementation of `gammanu`

Interesting, and what happened to water[False]?

Still net goodness for CPU-based CI, and we know IPU will require some additional work.

**Review by awf — APPROVED** [review]
